core/src/main/java/org/infinispan/configuration/parsing/SFSToSIFSStore.java

```diff
@@ -0,0 +1,108 @@
+ * that the new SoftIndexFileStore.
+ * @param <K> key type
+ * @param <V> value type
+ */
+@ConfiguredBy(SFSToSIFSConfiguration.class)
+public class SFSToSIFSStore<K, V> extends DelegatingNonBlockingStore<K, V> {
+   private NonBlockingStore<K, V> targetStore;
+
+   @Override
+   public NonBlockingStore<K, V> delegate() {
+      return targetStore;
+   }
+
+   @Override
+   public CompletionStage<Void> start(InitializationContext ctx) {
+      SFSToSIFSConfiguration sfsToSIFSConfiguration = ctx.getConfiguration();
+
+      CompletionStage<NonBlockingStore<K, V>> targetStage = createAndStartSoftIndexFileStore(ctx, sfsToSIFSConfiguration);
+
+      // If config has purge don't even bother with migration
+      if (sfsToSIFSConfiguration.purgeOnStartup()) {
+         return targetStage.thenAccept(sifsStore -> targetStore = sifsStore);
+      }
+
+      CompletionStage<NonBlockingStore<K, V>> nonSegmentedSFSStage = createAndStartSingleFileStore(ctx, sfsToSIFSConfiguration.dataLocation(), false);
+      CompletionStage<NonBlockingStore<K, V>> segmentedSFSStage = createAndStartSingleFileStore(ctx, sfsToSIFSConfiguration.dataLocation(), true);
+
+      Function<NonBlockingStore<K, V>, CompletionStage<Void>> composed = sourceStore -> targetStage.thenCompose(targetStore -> {
+         this.targetStore = targetStore;
+
+         int segmentCount = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
+         IntSet allSegments = IntSets.immutableRangeSet(segmentCount);
+
+         Flowable<GroupedFlowable<Integer, MarshallableEntry<K, V>>> groupedFlowable =
+               Flowable.fromPublisher(sourceStore.publishEntries(allSegments, null, true))
+                     .groupBy(ctx.getKeyPartitioner()::getSegment);
+
+         return targetStore.batch(segmentCount, Flowable.empty(), groupedFlowable.map(SegmentPublisherWrapper::wrap))
+               .thenCompose(ignore -> sourceStore.destroy());
+      });
+
+      return CompletionStages.allOf(nonSegmentedSFSStage.thenCompose(composed),
+            segmentedSFSStage.thenCompose(composed));
+   }
+
+   CompletionStage<NonBlockingStore<K, V>> createAndStartSingleFileStore(InitializationContext ctx, String location,
+         boolean segmented) {
+      AttributeSet sfsAttributes = SingleFileStoreConfiguration.attributeDefinitionSet();
+      sfsAttributes.attribute(SingleFileStoreConfiguration.LOCATION).set(location);
+      sfsAttributes.attribute(AbstractStoreConfiguration.SEGMENTED).set(segmented);
+      sfsAttributes.attribute(AbstractStoreConfiguration.IGNORE_MODIFICATIONS).set(Boolean.TRUE);
+
+      return createAndStartStore(ctx, new SingleFileStoreConfiguration(sfsAttributes.protect(),
+            new AsyncStoreConfiguration(AsyncStoreConfiguration.attributeDefinitionSet().protect())));
+   }
+
+   CompletionStage<NonBlockingStore<K, V>> createAndStartSoftIndexFileStore(InitializationContext ctx,
+         SFSToSIFSConfiguration sfsToSIFSConfiguration) {
```

Comment:
`/sfsToSIFSConfiguration/config` would make this much easier to read.

Reply:
Alright